design/design.md

```diff
@@ -0,0 +1,203 @@
+      image
+* All importers in a single repo
+* Importer authors own PR reviews
+* Requires a redeploy of the event image to update importers
+* Increases risk of malicious code injection over option 1
+
+### 3. Add external repos per importer with implementation code at build time
+* Isolates the event processor from the importer code
+* Isolates various importers from each other
+* Multiple dependency files (pipenv / uv / ...)
+    * Install and zip each module's dependencies separately at build time
+    * Provide the correct dependency zip for each job
+* Difficult to run integration tests on the importer code
+    * May be possible by cloning the event repo in GHA and building the event processor
+      image
+* Importer authors own PR reviews
+* Requires a redeploy of the event image to update importers
+* Increases risk of malicious code injection over option 2
+
+### 4. Dynamically update importers from external repo(s)
+* As option 2 or 3, but updates repo dynamically without requiring an image rebuild
+    * Check would occur as part of event loop
+* May need a location to dynamically pull configuration in order to add new importer repos without
+  redeployment
+    * Perhaps a configuration file in S3
+* Increases risk of malicious code injection over option 3
+
+## Implications for the CTS
+
+* We may need to have a set of topics for jobs to separate jobs that are expected to 
+  require extremely large processing times from more reasonable jobs, as those jobs will
+  block the rest of the topic partition from processing.
+  * Alternatively long jobs could be detected by the event processor and written to a different
+    topic.
+* In some cases multiple jobs may need to finish before the event processor can import the data,
+  if the data is related in such a way that importing one job's data doesn't make sense.
+  The CTS may need something like job sets to which the event processor can respond.
+* In some cases a string of jobs may need to complete. Potentially the event manager could
+  kick off another job based on the completed job.
+* This is reinventing the workflow manager wheel...
+
+## Requirements for importer authors
+
+* Importer authors making use of the CDM events processor must implement importers expecting that
+  the same event may be processed multiple times. The CTS only guarantees at-least-once
+  event processing.
+    * The [MERGE](https://docs.databricks.com/gcp/en/delta/merge) SQL statement
+      is one way to ensure duplicate data is not entered into the CDM.
+* Importer authors must ensure their code covers different versions of registered CTS images.
+  The code should key off the image digest if processing differences are required between different
+  image versions.
+
+## Out of scope
+
+* Monitoring the event processor
+    * At some point we will want to get alerts if the event processor goes down
+* Supporting multiple python versions
+    * Only support 3.11 (or .12?) for now
```

Comment:
uv can install and manage its own python - not sure if pipenv can do this?

Reply:
It can't other than erroring if the specified python version isn't installed AFAIK, but since the code is going to be running on the spark workers, UV being able to install python doesn't help